### DISCUSSION TITLE
Does brew has any plan to trim formulae's dependency trees?

### DISCUSSION THREAD

**Discussion by ivaquero** [Homebrew/discussions#5167] [Everyday usage] [answered]
For some formulae, I find they has heavier dependencies than it does in other pkg manager, especilly for c++ pkgs. For example

- in armadillo.rb, arpack should not be a dependency
- in matplotplusplus.rb, it requires qt which is very heavy and not a must for the formula

Does Homebrew team has any plan to reduce the dependencies?

**Comment by SMillerDev** [comment]
> For some formulae, I find they has heavier dependencies than it does in other pkg manager

Homebrew is also different from other package managers 

> in armadillo.rb, arpack should not be a dependency

Why not?
> in matplotplusplus.rb, it requires qt which is very heavy and not a must for the formula

But it is useful for the person who suggested adding it.
> Does Homebrew team has any plan to reduce the dependencies?

No, but you can make a pull request at any time to suggest changes and someone will happily review it.

**Comment by alebcay** [comment, marked as answer]
> I find they has heavier dependencies than it does in other pkg manager, especilly for c++ pkgs

Ultimately, this boils down to a design decision in (modern) Homebrew: packages in homebrew-core may not have optional dependencies. The two primary reasons for this choice, to my understanding, are:

- Everything in homebrew-core is distributed as bottles (precompiled binaries for each target platform). There ultimately needs to be a decision on what dependencies this precompiled package has, especially with a language like C++.
  - It's not feasible for Homebrew to build and maintain a copy of precompiled packages for each optional dependency. Naively, for a formula that has _n_ optional dependencies, if Homebrew were to offer precompiled packages for all possible permutations of optional dependencies to be included, there would need to be _2^n_ precompiled versions of the same package.
- Offering support to users similarly becomes more difficult as there would be _2^n_ possible configurations for a given package instead of one. This makes testing, troubleshooting, bisecting issues, reproducing problems, etc. much harder. When one also considers that the rate of new packages being added into homebrew-core is greatly outpacing the number of new maintainers, the thought of offering support in this manner becomes more untenable.

Therefore, the two paths to consider are (1) have more complete, functional packages that may carry bloat, or (2) have more minimal, barebones packages that may not have the features that most(?) users want. Between these two possible paths, Homebrew tends to lean towards the former.

---

Note that in third-party taps, formulae _may_ add optional dependencies. However, to take advantage of this, the packages must then be built from source on the end user's local machine (losing the benefit of bottles/precompiled packages).

**Reply by ivaquero** [reply]
Thank you for the detailed explanation
